drivers/sftp/onitu_sftp/sftp.py

```diff
@@ -0,0 +1,325 @@
+    finally:
+        if error is not None:
+            raise DriverError(u"Failed to create private key: {}"
+                              .format(error))
+    return private_key
+
+
+def get_SFTP_client(hostname, port, username, password,
+                    pkey_path, passphrase):
+    """Helper function to connect with SFTP. Tries to connect with the given
+    connection infos and, if successful, returns a SFTPClient ready to be
+    used.
+    Raises DriverError if an error occurred during the client setup."""
+    transport = paramiko.Transport((hostname, port))
+    try:
+        if password != '':
+            transport.connect(username=username, password=password)
+        else:
+            pkey_path = os.path.expanduser(pkey_path)
+            privateKey = get_private_key(pkey_path, passphrase)
+            transport.connect(username=username, pkey=privateKey)
+    except SSHException as se:
+        raise DriverError(u"Failed to connect to host: {}".format(se))
+    sftp = paramiko.SFTPClient.from_transport(transport)
+    return sftp
+
+
+def get_SFTP_client_from_plug():
+    """Helper function to connect an SFTP client with the plug infos."""
+    return get_SFTP_client(plug.options['hostname'],
+                           plug.options['port'],
+                           plug.options['username'],
+                           plug.options['password'],
+                           plug.options['private_key_path'],
+                           plug.options['private_key_passphrase'])
+
+
+def create_file_subdirs(sftp, filepath):
+    # Getting rid of the filename
+    lastSlash = filepath.rindex(u"/")
+    finalPath = filepath[:lastSlash]
+    dirs = finalPath.split(u"/")
+
+    currentDir = u""
+    doStat = True
+    try:
+        while dirs:
+            currentDir += dirs.pop(0) + u"/"
+            if doStat:
+                try:
+                    sftp.stat(currentDir)
+                # The current dir doesn't exists, so we create it
+                except IOError as ioe:
+                    if ioe.errno == 2:
+                        sftp.mkdir(currentDir)
+                        # If a dir was not found, it means all the next
+                        # won't exist either. So stop bothering to stat
+                        doStat = False
+                    else:
+                        raise
+            else:
+                sftp.mkdir(currentDir)
+    except IOError as ioe:
+        raise ServiceError(u"Error while creating subdirs {}: {}"
+                           .format(finalPath, ioe))
+
+
+def update_metadata(metadata, size, mtime, update_file=False):
+    metadata.size = size
+    metadata.extra['mtime'] = mtime
+    metadata.write()
+    if update_file:
+        plug.update_file(metadata)
+
+
+@plug.handler()
+def get_file(metadata):
+    plug.logger.debug(u"Getting remote file {}".format(metadata.path))
+    sftp = get_SFTP_client_from_plug()
```

Comment:
Not a big fan of opening a connection for each handler. Opening an SSH connection is very expensive (can be several seconds), so doing it at least three time for an upload is a pretty bad idea.
I think we should try to use only one with some locking for thread-safety. Regarding fault-tolerance, it really depends on how Paramiko handles it.